### DISCUSSION TITLE
How to store key and secret in a secure way

### DISCUSSION THREAD

**Discussion by matijahu1** [btschwertfeger/python-kraken-sdk#188] [Q&A]
Regarding:
```
key = os.getenv("API_KEY")
secret = os.getenv("SECRET_KEY")
```
I hesitate to store key and secret in the environment variables unencrypted. Isn't that a security risk? Is there a better way?

**Comment by matijahu1** [comment]
I thought a bit about that and had the idea that I can encrypt the password myself and only store the encrypted password.

**Comment by btschwertfeger** [comment]
Storing credentials securely is indeed a crucial aspect of application development, and there are various approaches to achieve this. Let's delve into the pros and cons of using environment variables for storing keys and secrets, and explore alternative methods.

### Storing Credentials in Environment Variables
#### Pros
**Convenience**: Environment variables provide a convenient way to pass sensitive information to applications without hardcoding them into the codebase.
**Security**: Environment variables are isolated from the application code and configuration files, reducing the risk of accidental exposure.
**Compatibility**: Environment variables are widely supported across different platforms and deployment environments, such as Docker containers and Kubernetes clusters.

#### Cons
**Limited Encryption**: Environment variables themselves are not encrypted, so sensitive data stored in them is exposed to anyone with access to the system environment.
**Potential Leakage**: If an attacker gains access to the system, they can easily retrieve environment variables containing sensitive information.
**Visibility**: Environment variables may be visible to users with sufficient permissions, making it challenging to control access to sensitive data.

### Alternative Approaches
1. **Encrypted Environment Variables**:
You can encrypt sensitive data before storing it as environment variables and decrypt it at runtime when needed. However, this approach introduces complexity and requires secure key management practices.

2. **Secret Management Tools**:
Utilize specialized secret management tools like HashiCorp Vault or AWS Secrets Manager to securely store and retrieve sensitive information. These tools offer features such as encryption at rest, access control, and audit logging.

3. **Configuration Files**:
Store credentials in encrypted configuration files and decrypt them at runtime. This approach provides more control over encryption and access policies but may require additional setup and management.

### Conclusion
While storing credentials as environment variables offers convenience and compatibility, it's essential to consider the security implications. Depending on the sensitivity of the data and the deployment environment, alternative approaches like encrypted environment variables or dedicated secret management tools may provide better protection against unauthorized access. Ultimately, the choice depends on balancing security requirements with operational convenience.

**Comment by adamantmm** [comment]
For anything beyond a local experiment I avoid putting keys in plaintext env files on disk.

What has worked for me with this SDK:

- load key/secret from the OS keychain (or a secrets manager) at process start, inject into the client, never write them back out
- restrict the API key to trade-only permissions and a fixed IP allowlist when the host is stable
- keep a separate read-only key for monitoring/scripts so a leaked bot key cannot withdraw

If you are on a VPS without a keychain, even a root-owned file mode `600` read once into memory beats exporting `KRAKEN_SECRET` in shell history.
